### ISSUE TITLE
feature suggestion: posterior probability of being in the zero-inflation state

### ISSUE THREAD

**Issue by jsocolar** [paul-buerkner/brms#1551]
I wrote the following function to predict the posterior probability of being in the zero-inflation state for the discrete zero-inflated families in `brms`.  If of interest, I'll submit a PR with doc + tests.

```r
posterior_zi_prob <- function (brmsfit, ...) {
  disc_zi_families <- c(
    "zero_inflated_poisson", "zero_inflated_negbinomial",
    "zero_inflated_binomial", "zero_inflated_beta_binomial"
  )
  
  fam <- brmsfit$family$family
  
  if(!(fam %in% disc_zi_families)) {
    stop2(paste0(
      "zi_probs available only for discrete zero-inflated families (",
      paste(disc_zi_families, collapse = ", "),
      "). Supplied brmsfit has family: ",
      fam
    ))
  }
  
  lik <- exp(log_lik(brmsfit, ...))
  zi_lik_given_obs_zero <- posterior_linpred(
    brmsfit, 
    dpar = "zi",
    transform = TRUE,
    ...
  )
  
  resp_name <- brmsfit$formula$resp
  resp_data <- brmsfit$data[[resp_name]]
  resp_is_zero <- as.integer(resp_data == 0)
  
  zi_lik <- sweep(zi_lik_given_obs_zero, MARGIN=2, resp_is_zero, `*`)
  
  zi_lik / lik
}

```

**Comment by paul-buerkner** [comment]
I am a bit confused. What is the difference of your suggestion to just predicting zi directly via `posterior_epred(..., dpar = "zi")`? After all zi is by definition the probability of being in the zero-inflation arm.

**Comment by jsocolar** [comment]
This refers to a situation where we care about the posterior probability that a particular observation arises from the latent zero-inflation state.

This probability of being in the zero-inflation state is zero given an observed nonzero, and is higher than `zi` given an observed zero.

**Comment by paul-buerkner** [comment]
I don't fully understand the when this is needed. So posterior_epred(dpar = "zi") give the posterior predicted probability for any (new) observation given predictors. You make it sound as if we now additionally condition the predcictions on (new?) observed data. Or am I misunderstanding something.

**Comment by jsocolar** [comment]
Just a note on naming: I guess `posterior_zi_prob()` is confusing and not well differentiated from `posterior_linpred(dpar = "zi", transform = T)` and `posterior_epred(..., dpar = "zi")`.  Maybe something like `latent_zi_state()` would be better.

**Comment by paul-buerkner** [comment]
Thank you for explaining. So this is basically getting the conditional mixture weights per observation? If so, perhaps we can generalize pp_mixture to not only work with formal mixture models (created via `mixture`) but also with de-facto mixture models such as zero-inflated and hurdle models? This would, I think, lead to a more consistent interface.

**Comment by jsocolar** [comment]
Sounds good. Note that in hurdle models the mixture weights are always either 0 or 1 and depend only on data, with no dependency on parameters. Do you think it's worthwhile to enable `pp_mixture` to compute the trivial mixture weights for hurdle models?  For zero-inflated models, only a subset of the weights are trivial.

**Comment by paul-buerkner** [comment]
Yeah, that make sense. For consistency, it could still be nice to have the hurdle models still even though the answer is trivial.

**Comment by jsocolar** [comment]
Edge case: what is the desired return of `pp_mixture` if a zero-inflated or hurdle (or zero-one-inflated) model is included as a mixture component in an explicit mixture model, e.g.

```
set.seed(1)
dat <- data.frame(
  y = runif(100)
)
mix2 <- mixture(gaussian, zero_one_inflated_beta)
fit2 <- brm(bf(y ~ 1), dat2, family = mix2) 
```

I tried checking if the behavior is already defined for nested mixtures in `brms`, but it appears not to be, as this errors:
```
mix3 <- mixture(mixture(gaussian, zero_one_inflated_beta), lognormal)
fit3 <- brm(bf(y~1), dat2, family = mix3)
```

My instinct is not to return the zero-inflation mixture weights at all and just return the weights of the explicitly defined (via `mixture()`) mixture components, and I note that any other choice would substantially change the existing behavior of `pp_mixture(fit3)`, rather than merely introduce new functionality.

**Comment by paul-buerkner** [comment]
I think we should have an argument that needs to be turned on explicitely for zi/hu models to be used directly within pp_mixture. This way we ensure that the functions behavior does not change too much just based on the input without explicit user intervention.
